### PR TITLE
[diffusion][feature] Add LingBot-Video distributed runtime support

### PR DESCRIPTION
**Pull request by wtz2333**
# [diffusion][feature] Add LingBot-Video distributed runtime support

## Summary

This PR extends the existing LingBot-Video integration with a unified multi-GPU runtime for the Dense 1.3B and MoE 30B-A3B Base checkpoints.

It adds:

- Ulysses sequence parallelism with a single joint-sequence shard/gather boundary;
- two-way CFG Parallel for Base and optional Refiner denoising;
- tensor parallel attention, Dense FFN, shared experts, and routed experts;
- vLLM common BF16 `FusedMoE` with load-time expert packing;
- serial VAE tiling and distributed VAE Patch Parallel decode;
- Cache-DiT compatibility with TP, Ulysses SP, CFG Parallel, and VAE Patch Parallel;
- a compile-friendly LingBot Transformer hot path;
- offline, serving, unit, CUDA, and weekly coverage for these runtime paths.

The public T2V, T2I, and TI2V request contracts introduced by #5035 and #5311 remain unchanged. The implementation continues to use the shared text-to-video runner from #6076 and the runtime foundation from #5581.

## Motivation

LingBot T2V latency is dominated by two components:

1. the long joint `[video; text]` sequence processed by every DiT block;
2. the positive and negative classifier-free guidance passes.

The MoE checkpoint also has a large resident-weight footprint and previously carried a model-specific grouped-matmul expert runtime. The VAE decode can become a separate memory and latency bottleneck for longer videos.

This PR addresses these paths through orthogonal runtime features:

- SP shards the joint sequence;
- CFG Parallel separates the two guidance branches;
- TP shards projections and expert computation;
- FusedMoE removes the duplicated expert runtime;
- VAE tiling and VAE Patch Parallel control decode memory and distribute tile work;
- Cache-DiT reduces repeated DiT computation and now composes with the supported parallel modes.

### Transformer runtime and BF16 FusedMoE

- Converts the LingBot Transformer to a native `torch.nn.Module`.
- Removes `text_lens.detach().cpu().tolist()` and per-sample Python construction from the denoising hot path.
- Vectorizes position, RoPE, text, and timestep metadata construction.
- Keeps attention, projections, normalization, AdaLN, and shared experts eligible for regional `torch.compile`.
- Replaces the model-specific routed-expert implementation with vLLM's common unquantized `FusedMoE`.
- Packs checkpoint `w1` and `w3` into `w13` during loading and loads `w2` directly.
- Preserves LingBot routing semantics, including FP32 router weights, correction bias, sigmoid scoring, group-limited top-k selection, and optional shared experts.

### Ulysses sequence parallelism

LingBot now exposes one model-level SP boundary:

```text
joint [video; text]
  -> build aligned RoPE, timestep, and validity metadata
  -> pad and shard once
  -> run the complete block stack with Ulysses attention
  -> gather once
  -> unpad and restore video tokens
```

This keeps the joint sequence and all aligned metadata on the same shard. Packed-CFG and text masks remain valid after resharding, with the common SDPA fallback selected when a fused backend cannot represent the required mask.

Ring sequence parallelism is rejected before sharding. It was tested and removed because it introduced additional constraints and numerical divergence without a material LingBot latency benefit.

### CFG Parallel

- Reuses the shared `CFGParallelMixin`.
- Supports CFG world sizes 1 and 2.
- Runs positive and negative branches on separate ranks and combines their predictions before each scheduler step.
- Shares the same denoising primitive between Base and optional Refiner.
- Synchronizes omitted request seeds across the SP and CFG groups.
- Rejects `batch_cfg=true` when CFG Parallel already owns the positive/negative split.
- Allows CFG-disabled engine warmup without requiring a negative branch.

SP and CFG compose. For example, SP2-CFG2 uses four workers as two independent SP groups, one per CFG branch.

### Tensor parallelism

- Uses column-parallel Q/K/V and MLP input projections.
- Uses row-parallel attention and MLP output projections.
- Loads checkpoint shards directly into local TP parameters.
- Applies TP to Dense FFN, shared experts, and routed experts.
- Passes the actual TP size to common BF16 FusedMoE while preserving checkpoint prefix mapping.
- Keeps TP1 behavior and the existing state-dict contract unchanged.

### VAE execution

Serial VAE tiling:

- uses the existing `vae_use_tiling` engine option;
- validates LingBot tile and stride geometry;
- accepts optional Base overrides through `model_config.lingbot_vae_tiling`;
- remains opt-in and preserves the non-tiled path.

Distributed VAE Patch Parallel:

- uses the shared `DistributedAutoencoderKLWan`;
- distributes tiled decode work across an existing DiT worker group;
- reconstructs the full video on DiT rank 0 without broadcasting it to every worker;
- preserves the empty non-primary-rank output sentinel;
- automatically enables tiling when VAE Patch Parallel is requested.

`vae_patch_parallel_size` does not create additional workers. It must reuse an existing DiT world of sufficient size.

### Cache-DiT with parallelism

Cache-DiT is supported with:

- tensor parallelism;
- Ulysses sequence parallelism;
- CFG Parallel;
- VAE Patch Parallel;
- ordinary CPU or layerwise offload.

For CFG1, Cache-DiT accounts for the two sequential cond/uncond forwards. For CFG2, each rank executes one branch and uses the single-forward cache contract.

Cache-DiT still rejects Pipeline Parallel, Expert Parallel, HSDP, and distributed layerwise offload combinations before model loading.

## Scope

| Feature | Dense Base | MoE Base | Validation in this PR |
| --- | --- | --- | --- |
| Ulysses SP1/SP2/SP4 | Yes | Yes | Standard T2V performance and raw-frame accuracy matrix |
| CFG1/CFG2 | Yes | Yes | Standard T2V performance matrix |
| SP × CFG composition | Yes | Yes | Through SP4-CFG2 on eight GPUs |
| Tensor Parallel TP1/TP2 | Yes | Yes | TP2 functional T2V smoke |
| Common BF16 FusedMoE | N/A | Yes | Eager/compile CUDA reference and multi-GPU T2V |
| Serial VAE tiling | Yes | Yes | Standard T2V matrix |
| VAE Patch Parallel | Yes | Yes | Dense TP2 + VAE-PP2 functional smoke |
| Cache-DiT + TP/SP/CFG/VAE-PP | Yes | Yes | Focused Dense functional smoke |
| Ring | No | No | Explicitly rejected |
| Expert Parallel | No | No | Out of scope |

The standard distributed benchmark covers Base T2V. Existing T2I, TI2V, and optional Refiner behavior is preserved, but this PR does not claim a full distributed performance or accuracy matrix for those modes.

## Usage

The following example uses SP2 and CFG2 on four GPUs:

```bash
python examples/offline_inference/text_to_video/text_to_video.py \
  --model robbyant/lingbot-video-dense-1.3b \
  --prompt-json /path/to/lingbot-video/assets/cases/t2v/example_1/prompt.json \
  --height 480 --width 832 --num-frames 121 \
  --num-inference-steps 40 --guidance-scale 3.0 --flow-shift 3.0 \
  --seed 42 --fps 24 \
  --ulysses-degree 2 --cfg-parallel-size 2 \
  --vae-use-tiling \
  --output lingbot_sp2_cfg2.mp4
```

Other feature flags can be selected independently:

- TP2: `--tensor-parallel-size 2`
- VAE Patch Parallel 2: `--vae-patch-parallel-size 2 --vae-use-tiling`
- Cache-DiT: `--cache-backend cache_dit`

### Environment

- Python 3.12
- PyTorch `2.13.0+cu130`
- vLLM `0.27.0`
- NVIDIA L20X, up to eight GPUs
- FlashAttention
- common unquantized Triton FusedMoE
- local Dense 1.3B and MoE 30B-A3B checkpoints
- regional `torch.compile` enabled

The current L20X runtime has no tuned FusedMoE configuration for this exact LingBot expert shape. MoE measurements are therefore functional deployment baselines rather than a tuned-kernel ceiling.

### Automated regression

| Validation | Result |
| --- | --- |
| Complete LingBot model test suite after the final TP/VAE/Cache-DiT commits | `204 passed` |
| Common FusedMoE eager/compile CUDA reference | `2 passed` |
| Ruff check and format | Passed |
| `git diff --check` | Passed |
| Weekly four-GPU SP2-CFG2 serving coverage | Added |

The final CPU suite was run with:

```bash
python -m pytest -q tests/diffusion/models/lingbot_video
```

### Functional multi-GPU validation

These checks use the official `example_1` prompt, 480x832, 49 frames, 40 denoise steps, guidance 3, shift 3, seed 42, 24 FPS, and regional compile.

They are functional smoke tests, not performance benchmarks. Every row generated a valid 49-frame, 832x480, 24 FPS MP4.

| Model | Topology | Path validated | Result |
| --- | --- | --- | --- |
| Dense | TP2 | TP attention and Dense FFN | Passed |
| MoE | TP2 | TP attention and TP-aware FusedMoE | Passed |
| Dense | Cache-DiT + TP2 | Distributed cache decisions with TP | Passed |
| Dense | Cache-DiT + TP2 + VAE-PP2 | Distributed tiled VAE decode | Passed |
| Dense | Cache-DiT + SP2 | Distributed cache decisions with Ulysses | Passed |
| Dense | Cache-DiT + CFG2 | Per-rank single-branch cache contract (`CFG0`) | Passed |

### Standard Base T2V performance

This is the only performance matrix in the PR.

Workload:

- official `example_1` structured prompt;
- 480x832;
- 121 frames;
- 40 denoise steps;
- guidance 3;
- flow shift 3;
- seed 42;
- 24 FPS;
- serial VAE tiling;
- regional compile.

Each topology runs three requests in one resident engine. The first compile/warmup request is excluded and the final two are averaged. E2E includes condition handling, denoising, VAE decode, and MP4 export, but excludes model loading.

| Model | Topology | GPUs | Mean E2E | Speedup vs SP1-CFG1 | Aggregate peak process memory |
| --- | --- | ---: | ---: | ---: | ---: |
| Dense | SP1-CFG1 | 1 | 88.588 s | 1.000x | 22.586 GiB |
| Dense | SP1-CFG2 | 2 | 50.241 s | 1.763x | 46.171 GiB |
| Dense | SP2-CFG1 | 2 | 55.283 s | 1.602x | 43.874 GiB |
| Dense | SP2-CFG2 | 4 | 33.662 s | 2.632x | 87.054 GiB |
| Dense | SP4-CFG1 | 4 | 34.408 s | 2.575x | 82.003 GiB |
| Dense | SP4-CFG2 | 8 | 22.348 s | 3.964x | 171.266 GiB |
| MoE | SP1-CFG1 | 1 | 184.864 s | 1.000x | 76.424 GiB |
| MoE | SP1-CFG2 | 2 | 99.203 s | 1.863x | 153.663 GiB |
| MoE | SP2-CFG1 | 2 | 109.965 s | 1.681x | 155.023 GiB |
| MoE | SP2-CFG2 | 4 | 61.975 s | 2.983x | 304.159 GiB |
| MoE | SP4-CFG1 | 4 | 63.364 s | 2.917x | 299.480 GiB |
| MoE | SP4-CFG2 | 8 | 36.959 s | 5.002x | 593.415 GiB |

On two GPUs, CFG2 is approximately 9% faster than SP2 for both checkpoints. On four GPUs, SP2-CFG2 is slightly faster than SP4. SP4-CFG2 completes successfully for both models.

### SP raw-frame accuracy

Every SP result is compared with SP1 at the same CFG degree using `[0,1]` float frames retained before MP4 encoding.

| Model | Comparison | PSNR | MAE |
| --- | --- | ---: | ---: |
| Dense | SP2-CFG1 vs SP1-CFG1 | 34.578 dB | 0.007434 |
| Dense | SP2-CFG2 vs SP1-CFG2 | 35.184 dB | 0.007324 |
| Dense | SP4-CFG1 vs SP1-CFG1 | 32.995 dB | 0.009799 |
| Dense | SP4-CFG2 vs SP1-CFG2 | 32.504 dB | 0.010427 |
| MoE | SP2-CFG1 vs SP1-CFG1 | 18.305 dB | 0.071144 |
| MoE | SP2-CFG2 vs SP1-CFG2 | 18.446 dB | 0.071828 |
| MoE | SP4-CFG1 vs SP1-CFG1 | 18.305 dB | 0.071144 |
| MoE | SP4-CFG2 vs SP1-CFG2 | 18.305 dB | 0.071144 |

The Dense checkpoint remains above 32 dB throughout the matrix. The MoE matrix validates reproducible distributed execution but does not establish SP numerical parity on the current vLLM 0.27 runtime.

## Limitations

- MoE SP raw-frame parity remains below the intended threshold on the current runtime.
- Ring sequence parallelism is intentionally unsupported; use Ulysses with `ring_degree=1`.
- Cache-DiT is not combined with Pipeline Parallel, Expert Parallel, HSDP, or distributed layerwise offload.
- VAE Patch Parallel reuses the existing DiT world and does not create workers.
- Only VAE decode is distributed; image/video condition encoding remains on the existing encode path.
- T2I, TI2V, and optional Refiner do not have the full Base T2V distributed benchmark matrix.
- Expert-only FP8, Expert Parallel, and CUDA Graph remain follow-up work.

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
This PR touches **vllm_omni/diffusion/, tests/diffusion/, examples/offline_inference/** (8 files). Based on CODEOWNERS coverage of the changed files, the most-related reviewers appear to be:

> @Isotr0py @princepride @SamitHuang

Could one of you take a look when you get a chance? Thanks!

**Comment by wtz2333** [comment]
> This PR touches **vllm_omni/diffusion/, tests/diffusion/, examples/offline_inference/** (8 files). Based on CODEOWNERS coverage of the changed files, the most-related reviewers appear to be:此 PR 修改了 **vllm_omni/diffusion/、tests/diffusion/ 和 examples/offline_inference/ 目录**下的 8 个文件。根据 CODEOWNERS 对已修改文件的覆盖情况，最相关的审阅者似乎是：
> 
> > @Isotr0py @princepride @SamitHuang
> 
> Could one of you take a look when you get a chance? Thanks!请问哪位有空可以帮忙看一下吗？谢谢！

this PR is not ready

**Comment by wtz2333** [comment]
I have conducted fairly detailed performance tests.  https://github.com/wtz2333/vllm-omni/issues/2#issuecomment-5310985440

**Comment by wtz2333** [comment]
This PR is ready

File: .buildkite/cuda/test-nightly.yml [hunk 1]
```diff
@@ -292,6 +292,12 @@ steps:
           - pytest -s -v tests/e2e/online_serving/test_wan22_expansion.py -k "i2v" -m "full_model and cuda" --run-level "full_model"
         mirror_hardwares: h100_2
 
+      - label: ":full_moon: Diffusion X2V · LingBot Video MoE Function Test"
```
Comment (yenuo26) [inline .buildkite/cuda/test-nightly.yml:295]: Perhaps put it in the weekly CI?

File: .buildkite/cuda/test-weekly.yml [hunk 2]
```diff
@@ -142,3 +142,13 @@ steps:
             tests/e2e/ -m "slow and diffusion and H100 and distributed_cuda"
             --run-level "full_model"
         mirror_hardwares: h100_4
+
+      - label: "Diffusion · H100 · LingBot SP + CFG"
+        timeout_in_minutes: 90
+        commands:
+          - >-
+            pytest -s -v
+            tests/e2e/online_serving/test_lingbot_video_expansion.py
```
Comment (yenuo26) [inline .buildkite/cuda/test-weekly.yml:151]: Perhaps it should uniformly use the 'slow' tag. Since the weekly CI job will collect all tests with the 'slow' tag, if this needs to be a separate job, we'll need to exclude it from the other diffusion jobs, otherwise it will run twice.
